### PR TITLE
Issue #7040: spotbugs is not executed in build process

### PR DESCRIPTION
**Pull request by krichter722**
#7040

Add exclusions to spotbugs-exclude.xml for Files.newInputStream which can be ignored

I don't know why this doesn't fail on CI, but it blocks usage of `mvn clean install` locally. The messages don't reveal any issues and can be ignored.

### CODE REVIEW

**Comment by rnveach** [comment]
I assume this is another JDK12 issue.
We skip and a few things in the JDK12 run: https://github.com/checkstyle/checkstyle/blob/master/.ci/appveyor.bat#L29

**Comment by krichter722** [comment]
@rnveach Does that mean that you're in favour or opposed to this change? The issues occur on all of OpenJDK 8, 11 and 12 when I run `mvn install` locally.

**Comment by rnveach** [comment]
If we make this change, you have to remove the exclusion for spotbugs to show the change fixes it in the CI.
I do question making an exclusion for an entire method when a JDK version far beyond what we currently use requires it for the whole project. I was waiting for @romani to give his opinion.

**Comment by krichter722** [comment]
Judging from the CI output of master and latest commit I pushed here spotbugs was never running for OpenJDK < 12. I suggest adding it for OpenJDK 8 and 12 which are GA and OpenJDK 11 which has only recently be deprecated.

**Comment by rnveach** [comment]
> spotbugs was never running for OpenJDK < 12. I suggest adding it for OpenJDK 8 and 12 which are GA and OpenJDK 11 which has only recently be deprecated.

We do run it for openjdk 8.
See https://github.com/checkstyle/checkstyle/blob/master/.travis.yml#L55-L58

**Comment by romani** [comment]
@krichter722 , please define a problem and what problem you experience. 

You can create issue on it or you can treat this PR as issue and reference in commit it's number "Pull #7034: ...".
Minor prefix is for stuff that is not a problem and just polishing of something.

**Comment by romani** [comment]
> but it blocks usage of mvn clean install locally.

Is it resolved ? if yes, lets make CI to control it.

> `spotbugs:check -Dspotbugs.skip=false`

please explain this hack

File: .travis.yml [hunk 1]
```diff
@@ -55,7 +55,21 @@ matrix:
     - jdk: openjdk8
       env:
         - DESC="spotbugs,pmd"
-        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean compile pmd:check spotbugs:check"
+        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean compile pmd:check spotbugs:check -Dspotbugs.skip=false"
```
Comment (romani) [inline .travis.yml]: no changes for openjdk is expected.

File: .travis.yml [hunk 2]
```diff
@@ -55,7 +55,21 @@ matrix:
...
+        - USE_MAVEN_REPO="true"
+
+    # spotbugs and pmd (openjdk11)
+    - jdk: openjdk11
+      env:
+        - DESC="spotbugs,pmd"
+        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean compile pmd:check spotbugs:check -Dspotbugs.skip=false"
```
Comment (romani) [inline .travis.yml]: please explain why `-Dspotbugs.skip=false` i required.

File: config/spotbugs-exclude.xml [hunk 3]
```diff
@@ -45,6 +45,46 @@
       <Method name="main" />
       <Bug pattern="DM_EXIT" />
     </Match>
+    <Match>
+        <Class name="com.puppycrawl.tools.checkstyle.Main" />
+        <Method name="loadProperties" />
+        <Bug pattern="RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE" />
```
Comment (romani) [inline config/spotbugs-exclude.xml]: please share in issue exact list of violations.

**Comment by rnveach** [comment]
@krichter722 ping

File: .travis.yml (hunk 2 above)
Comment (krichter722) [reply .travis.yml]: It seems to be unnecessary if the target is specified explicitly, thanks.

**Comment by krichter722** [comment]
This patch allows to run `mvn clean install` locally. Excluding the hole method from the check is fine afaik since it's a weakness of JDK design. I updated the commit message to include the PR number.

File: config/spotbugs-exclude.xml (hunk 3 above)
Comment (romani) [reply config/spotbugs-exclude.xml]: Done.

@krichter722 , please keep addressing each review point, I replied points for us mean that you are still in progress.

Comment (romani) [inline config/spotbugs-exclude.xml]: Each suppress have corresponding comment to explain reason, please provide.
If case is complicated, it is better to do wide explanation in issue comment and make xml comment ` see link for details `

Comment (romani) [reply config/spotbugs-exclude.xml]: it would speed up process, if you start replyinig as "done" for each review process.

**Comment by romani** [comment]
I can not validate this PR, because I got output of spotbugs from jdk11 and placed it to issue description, but line numbers are pointing to comments in code.
It is not that easy for me to get to non jdk8 environment, if somebody have it, please share spotbugs output again.

File: config/spotbugs-exclude.xml [hunk 4]
```diff
@@ -45,6 +45,54 @@
       <Method name="main" />
       <Bug pattern="DM_EXIT" />
     </Match>
+    <Match>
+        <!-- this is a JDK design weakness that we can do nothing about -->
```
Comment (romani) [inline config/spotbugs-exclude.xml]: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L423-L425

Why it is JDK weakness ?
it is problem of spotbugs tool, it should not report issues on this such code.
We need to report issue on spotbugs and suppress until issue on their side is resolved.

comment should be `till LINK_TO_ISSUE`

Comment (romani) [reply config/spotbugs-exclude.xml]: we already experienced this problem, here is our team member post - https://github.com/spotbugs/spotbugs/issues/259#issuecomment-445460892

Comment (romani) [reply config/spotbugs-exclude.xml]: comment should be `till https://github.com/spotbugs/spotbugs/issues/259`

Comment (romani) [reply config/spotbugs-exclude.xml]: done.

**Comment by krichter722** [comment]
@rnveach good finding. This doesn't seem to be fixed by spotbugs so soon since it's open since 2017. What do you think about adding the spotbugs issue URL to the comment explaining the exclusion and enabling spotbugs for OpenJDK 11 and 12? The issue that final fields can't be mock with OpenJDK 12 is addressed in #7035.

**Comment by rnveach** [comment]
@romani https://teamcity.jetbrains.com/viewLog.html?buildId=2558933&buildTypeId=Checkstyle_IdeaInspectionsPullRequest&tab=buildLog&_focus=3132
> The build Checkstyle::IDEA Inspections Pull Request #12381 {buildId=2558933} has been running for more than 50 minutes. Terminating...

**Comment by romani** [comment]
Relaunched,

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/blob/master/.ci/appveyor.bat#L29
Need to be updated to remove `Dspotbugs.skip=true`

**Comment by romani** [comment]
appveyor is updated, @rnveach , please review.

**Pull request merged**